packages/mars/src/values/future.ts

```diff
@@ -40,3 +40,18 @@ export class Future<T> {
     throw new Error('This will get overridden.')
   }
 }
+
+type FutureTuple<T> = T extends [infer V, ...infer Rest] ? [Future<V>, ...FutureTuple<Rest>] : []
+
+type FutureArray<T> = T extends Array<any> ? FutureTuple<T> : Future<T>[]
+type UnpackArray<T> = T extends FutureArray<infer U> ? U : never
+type ForceArray<T> = T extends Array<any> ? T : T[]
+
+export function flatMap<T extends FutureArray<any>, R>(
```

Comment:
Now I understand, but....
it's rather a special kind of `reduce` instead of `map` or `flatMap`.

Pls compare the semantics of 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce

Here it's reducing the entire array of futures into a single value of type R

So, could you propose a new name and add a little doc on what it does?

Reply:
Ok, let's call it reduce